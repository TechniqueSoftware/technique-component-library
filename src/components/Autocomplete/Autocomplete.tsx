import * as React from 'react';
import MuiAutocomplete, {
  AutocompleteProps as MuiAutocompleteProps,
} from '@material-ui/lab/Autocomplete';
import { TextField, TextFieldProps } from '../TextField/TextField';
import parse from 'autosuggest-highlight/parse';
import match from 'autosuggest-highlight/match';

export interface AutocompleteProps<OptionType> {
  id: string;
  inputValue?: string;
  loading: boolean;
  open: boolean;
  disabled?: boolean;
  textFieldProps?: TextFieldProps;
  inputWidth: number;
  noOptionsText?: string;
  options: OptionType[];
  onClose: () => void;
  getOptionLabel: (optionType: OptionType) => string;
  onOpen: (event: React.ChangeEvent<any>) => void;
  onChange?: (event: React.ChangeEvent<any>, value: NonNullable<OptionType>) => void;
  onInputChange?: (event: React.ChangeEvent<any>, value: string, reason: 'input' | 'reset' | 'clear') => void;
}

export default function Autocomplete<OptionType>(autocompleteProps: AutocompleteProps<OptionType>) {
  const {
    id,
    inputValue,
    open,
    loading,
    textFieldProps,
    inputWidth,
    onClose,
    onOpen,
    options,
    onInputChange,
    getOptionLabel,
    disabled,
    onChange,
    noOptionsText = 'No results'
  } = autocompleteProps;

  const muiAutocompleteProps: MuiAutocompleteProps<OptionType, undefined, undefined, undefined> = {
    id,
    inputValue,
    open,
    onOpen,
    onClose,
    options,
    loading,
    disabled,
    getOptionLabel,
    onInputChange,
    noOptionsText,
    style: { width: inputWidth },
    getOptionSelected: (option, value) => getOptionLabel(option) === getOptionLabel(value),
    renderInput: params => {
      return <TextField {...params} {...textFieldProps} />;
    },
    renderOption: (option, { inputValue }) => {
      const label = getOptionLabel(option);
      const matches = match(label, inputValue);
      const parts = parse(label, matches);
      return (
        <div>
          {parts.map((part, index) => (
            <span key={index} style={{ fontWeight: part.highlight ? 700 : 400 }}>
            {part.text}
          </span>
          ))}
        </div>
      );
    }
  };

  return (
    <MuiAutocomplete {...muiAutocompleteProps} onChange={onChange} />
  );
}
