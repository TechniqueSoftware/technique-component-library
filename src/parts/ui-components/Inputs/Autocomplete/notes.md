# AutoComplete
This component wraps our `TextField` component and passes in the renderInput props to apply the [AutoComplete](https://material-ui.com/components/AutoComplete/) styles and functionality.

## Usage
This component was created initially for asynchronous call with autocomplete options.

## JIRA
Related jira ticket: https://clubos.atlassian.net/browse/SM-519
