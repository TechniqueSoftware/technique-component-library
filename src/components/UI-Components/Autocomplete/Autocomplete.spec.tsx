import Autocomplete, { AutocompleteProps } from './Autocomplete';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('Autocomplete', () => {

  test('should render with base props', () => {
    const baseProps: AutocompleteProps<any> = {
      getOptionLabel: (option) => option.label,
      open: false,
      inputValue: '',
      loading: false,
      inputWidth: 300,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      options: [],
      id: 'asynchronous-demo',
    };
    const shallowWrapper = shallow(<Autocomplete {...baseProps} />);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });
});
