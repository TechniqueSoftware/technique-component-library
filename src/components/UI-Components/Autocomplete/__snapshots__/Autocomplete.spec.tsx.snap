// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render with base props 1`] = `"<WithStyles(ForwardRef(Autocomplete)) id=\\"asynchronous-demo\\" inputValue=\\"\\" open={false} onOpen={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} onClose={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} options={{...}} loading={false} disabled={[undefined]} getOptionLabel={[Function: getOptionLabel]} onInputChange={[undefined]} noOptionsText=\\"No results\\" style={{...}} getOptionSelected={[Function: getOptionSelected]} renderInput={[Function: renderInput]} renderOption={[Function: renderOption]} onChange={[undefined]} />"`;
